### PR TITLE
picture note

### PR DESCRIPTION
**Pull request by ggjj321**
Add camera page

### CODE REVIEW

**Comment by ggjj321** [comment]
takePicture() function leads to a crash on a virtual machine but doesn't occur on a real machine.
I think it is related to the controller.

**Comment by Xanonymous-GitHub** [comment]
Please resolve the conflicts and rebase to master, then I'll keep helping on debugging the photo-related bugs 😃.

**Comment by Xanonymous-GitHub** [comment]
Seems there has some style/lint/format errors after [3683abc](https://github.com/NEO-TAT/tat_flutter/pull/83/commits/3683abcdc0b19c49f6b811a9232a114094b33bf4).
You may use the following steps to find and fix these problems:
```sh
flutter format --fix -l 120 .
```

```sh
dart fix --apply
```

```sh
flutter analyze
```

Please make sure that the above commands are in normal output before making the next modification.

An example output is:
截圖 2022-05-13 上午6 36 41


And you might refer to these official docs & developer's medium posts(my simply recommended):
- https://dart.dev/guides/language/effective-dart
- https://docs.flutter.dev/development/tools/flutter-fix
- https://medium.com/@inficial/flutter-best-practices-for-improve-performance-7e21e14efebb

**Comment by Xanonymous-GitHub** [comment]
Hi @ggjj321 , please fix the conflicts of this PR, then we could start the review.

File: lib/src/store/picture_storage.dart [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import 'package:get/get.dart';
+import 'package:sqflite/sqflite.dart';
+import 'dart:io';
+
+class PictureStorage {
+  static pictureInformationMap(String courseId, String label, String picturePath, String note) {
```
Comment (ricky9667) [inline lib/src/store/picture_storage.dart:6]: Suggest declaring this static function with a return type.

```suggestion
  static Map<String, String> pictureInformationMap(String courseId, String label, String picturePath, String note) {
```

File: lib/src/store/picture_storage.dart [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    return {
+      'courseId': courseId,
+      'label': label,
+      'note': note,
+      'picturePath': picturePath,
+    };
+  }
+
+  static takePictureToStorage(String courseId, String picturePath) async {
```
Comment (ricky9667) [inline lib/src/store/picture_storage.dart:15]: Ditto.

```suggestion
  static Future<void> takePictureToStorage(String courseId, String picturePath) async {
```

File: lib/src/store/picture_storage.dart [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    Database pictureDB = Get.find<Database>();
```
Comment (ricky9667) [inline lib/src/store/picture_storage.dart:16]: Suggest adding final to this variable.

```suggestion
    final pictureDB = Get.find<Database>();
```

File: lib/src/store/picture_storage.dart [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    );
+  }
+
+  static getCoursePicture(String courseId) async {
+    Database pictureDB = Get.find<Database>();
+    List<Map> picturePaths =  await pictureDB.rawQuery('SELECT  * '
+                          'FROM photo_storage '
+                          'WHERE courseId=$courseId');
+    return picturePaths;
+  }
```
Comment (ricky9667) [inline lib/src/store/picture_storage.dart:31]: There are several problems here:
1. You should give this method a return type.
2. It may be better to specify the key and value type for `Map` for picture paths, it may be `Map<String, dynamic>` in my opinion.
3. The `return` can be combined with the previous line.
4. Check your formatting before you commit, you accidentally added an extra space before `await`.

After reviewing these, this method can be improved in this way:
```suggestion
  static Map<String, dynamic> getCoursePicture(String courseId) async {
    final pictureDB = Get.find<Database>();
    return await pictureDB.rawQuery('SELECT  * '
                          'FROM photo_storage '
                          'WHERE courseId=$courseId');
  }
```

File: lib/src/store/picture_storage.dart [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+class Picture{
+  late int id;
+  late String label;
+  late String note;
+  late String path;
```
Comment (ricky9667) [inline lib/src/store/picture_storage.dart:38]: It seems like `late` can be removed, not really sure what this is for.
Therefore, suggest adding `final` if the property will not be modified.

File: lib/src/store/picture_storage.dart [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+  Picture(this.id, this.label, this.note, this.path);
+
+  String getLabel() => label;
+
+  String getNote() => note;
+
+  String getPath() => path;
```
Comment (ricky9667) [inline lib/src/store/picture_storage.dart:46]: Suggest using getters in this case.
More importantly, if you want to implement encapsulation, suggest make the properties private (ex. `_label`).

```suggestion
  String getLabel() => label;

  String getNote() => note;

  String getPath() => path;
```

File: lib/src/store/picture_storage.dart [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+  void modifyLabel(){}
+
+  void modifyNote(){}
+
+  void deletePicture() async {
```
Comment (ricky9667) [inline lib/src/store/picture_storage.dart:52]: The return type of this method should be a `Future` in my opinion, since the function is asynchronous.

```suggestion
  Future<void> deletePicture() async {
```

File: lib/ui/other/route_utils.dart [hunk 8]
```diff
@@ -153,7 +154,14 @@ class RouteUtils {
           onPageClose: () => Get.back(),
         ),
         transition: transition,
+        preventDuplicates: true,
       );
 
   static bool isLoggedIntoZuvio() => ZAuthController.to.isLoggedIntoZuvio();
+
+  static Future<void> launchCameraPage(String courseId) => Get.to(
```
Comment (ricky9667) [inline lib/ui/other/route_utils.dart:162]: This `Future` seems unnecessary.

```suggestion
  static void launchCameraPage(String courseId) => Get.to(
```

File: lib/ui/pages/coursedetail/screen/album_page.dart [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+  final String courseId;
+
+  const AlbumPage({Key? key, required this.courseId}) : super(key: key);
+
+  @override
+  State<AlbumPage> createState() => _AlbumPageState();
+}
+
+class _AlbumPageState extends State<AlbumPage> {
+  bool isLoading = true;
```
Comment (ricky9667) [inline lib/ui/pages/coursedetail/screen/album_page.dart:16]: Suggest making this property private.

File: lib/ui/pages/coursedetail/screen/album_page.dart [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+  late int pictureNum;
```
Comment (ricky9667) [inline lib/ui/pages/coursedetail/screen/album_page.dart:18]: Declare this inside `useCourseIdToGetPicturePaths()`, it does not have other usages.

File: lib/ui/pages/coursedetail/screen/album_page.dart [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    isLoading = false;
+    pictureNum = picturesInfo.length;
+    for (int i = 0; i < pictureNum; i++) {
+      int infoId = picturesInfo[i]['_id'];
+      String infoLabel = picturesInfo[i]['label'];
+      String infoNote = picturesInfo[i]['note'];
+      String infoPath = picturesInfo[i]['picturePath'];
+
+      pictures.add(Picture(infoId, infoLabel, infoNote, infoPath));
+    }
```
Comment (ricky9667) [inline lib/ui/pages/coursedetail/screen/album_page.dart:45]: This can be simplified.

```suggestion
    for (final pictureInfo in pictureInfos) {
      int infoId = pictureInfo['_id'];
      String infoLabel = pictureInfo['label'];
      String infoNote = pictureInfo['note'];
      String infoPath = pictureInfo['picturePath'];

      pictures.add(Picture(infoId, infoLabel, infoNote, infoPath));
    }
```

File: lib/ui/pages/coursedetail/screen/album_page.dart [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+  State<AlbumPage> createState() => _AlbumPageState();
+}
+
+class _AlbumPageState extends State<AlbumPage> {
+  bool isLoading = true;
+
+  late int pictureNum;
+  late List<Picture> pictures = [];
+
+  dynamic picturesInfo;
```
Comment (ricky9667) [inline lib/ui/pages/coursedetail/screen/album_page.dart:21]: Declare this inside useCourseIdToGetPicturePaths(), it does not have other usages.

File: lib/ui/pages/coursedetail/screen/album_page.dart [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+      child: Column(
+        children: [
+          isLoading
+              ? const Center(
+                  child: CircularProgressIndicator(),
+                )
+              : Expanded(
+                  child: pictureAlbum(),
+                ),
+        ],
```
Comment (ricky9667) [inline lib/ui/pages/coursedetail/screen/album_page.dart:64]: There is only one child in `children`, suggest removing `Column`.

File: lib/ui/pages/coursedetail/screen/album_page.dart [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+      ),
+    );
+  }
+
+  Widget pictureAlbum() {
+    return ListView(children: [
```
Comment (ricky9667) [inline lib/ui/pages/coursedetail/screen/album_page.dart:70]: Use getter in this case.

```suggestion
  Widget get pictureAlbum {
    return ListView(children: [
```

File: lib/ui/pages/coursedetail/screen/album_page.dart [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+          },
+          label: const Text('Delete'),
+          backgroundColor: Colors.pink,
+        ),
+          floatingActionButtonLocation: FloatingActionButtonLocation.centerFloat
+      );
+    }));
+  }
+}
+//
```
Comment (ricky9667) [inline lib/ui/pages/coursedetail/screen/album_page.dart:115]: Please remove this comment.

File: lib/ui/pages/coursedetail/screen/album_page.dart [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+            pictures.remove(picture);
+            setState(() {});
+            Navigator.pop(context);
+          },
+          label: const Text('Delete'),
+          backgroundColor: Colors.pink,
+        ),
+          floatingActionButtonLocation: FloatingActionButtonLocation.centerFloat
+      );
+    }));
```
Comment (ricky9667) [inline lib/ui/pages/coursedetail/screen/album_page.dart:112]: Suggest extracting the page route before `push`.

```
final pageRoute = MaterialPageRoute(
    ...
);
Navigator.push(context, pageRoute);
```

File: lib/ui/pages/coursetable/course_table_page.dart [hunk 17]
```diff
@@ -695,6 +700,13 @@ class _CourseTablePageState extends State<CourseTablePage> {
     RouteUtils.launchRollCallDashBoardPageAfterLogin();
   }
 
+  Future<void> _showCameraPage(CourseInfoJson courseInfo) async {
+    CourseMainJson course = courseInfo.main.course;
+    String courseId = course.id;
+    Get.back();
+    RouteUtils.launchCameraPage(courseId);
```
Comment (ricky9667) [inline lib/ui/pages/coursetable/course_table_page.dart:707]: This can be simplified.

```suggestion
    Get.back();
    RouteUtils.launchCameraPage(course.id);
```

File: lib/ui/pages/note_camera/note_camera.dart [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+  const NoteCamera({Key? key, required this.courseId}) : super(key: key);
+
+  @override
+  State<NoteCamera> createState() => _NoteCameraState();
+}
+
+class _NoteCameraState extends State<NoteCamera> with WidgetsBindingObserver {
+  CameraController? controller;
+  bool _isCameraInitialized = false;
+  late List<CameraDescription> cameras;
```
Comment (ricky9667) [inline lib/ui/pages/note_camera/note_camera.dart:19]: Suggest assigning an empty list instead of relying on `late`.

```suggestion
  List<CameraDescription> cameras = const [];
```

File: lib/ui/pages/note_camera/note_camera.dart [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+    super.initState();
+    SystemChrome.setEnabledSystemUIMode(SystemUiMode.manual, overlays: []);
+
+    cameras = Get.find<List<CameraDescription>>();
+    onNewCameraSelected(cameras[0]);
+  }
+
+  @override
+  void didChangeAppLifecycleState(AppLifecycleState state) {
+    final CameraController? cameraController = controller;
```
Comment (ricky9667) [inline lib/ui/pages/note_camera/note_camera.dart:75]: Remove the type when you declare with `final`.

```suggestion
    final cameraController = controller;
```

File: lib/ui/pages/note_camera/note_camera.dart [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+        if (scaleStartDetails.scale > 1.0) {
+          newZoom = _zoom + 0.05;
+        } else if (scaleStartDetails.scale < 1.0) {
+          newZoom = _zoom - 0.05;
+        }
+        if (newZoom >= 1.0 && newZoom <= 9.0) {
+          _zoom = newZoom;
+        }
+        controller?.setZoomLevel(_zoom);
+      },
```
Comment (ricky9667) [inline lib/ui/pages/note_camera/note_camera.dart:109]: Suggest extracting this callback to a new private method.

File: lib/ui/pages/coursedetail/screen/album_page.dart (hunk 16 above)
Comment (ggjj321) [reply lib/ui/pages/coursedetail/screen/album_page.dart]: I will create another page file to solve it.

**Comment by ggjj321** [comment]
The package "camera" has a problem on Android 13. If my Android SDK version is 33, it will occur a problem. https://github.com/flutter/flutter/issues/109769
If my Android SDK version is 31, it can successfully work.

**Comment by ggjj321** [comment]
Conflicts and Ricky's suggestion have been completed.
It still has some problems.
Please add these for the IOS camera. https://pub.dev/packages/camera
image
flutter analyze problem
image

**Comment by Xanonymous-GitHub** [comment]
Hi @ggjj321 , thanks for your contribution. It seems that except for the iOS settings that need to be added and tested, the rest of the work has been roughly completed.
Regarding @ricky9667's review, since you didn't **cut the commit for each comment and shows you'd addressed them**, I may take a moment to check whether you have made proper corrections for each comment he made, and I also will review this PR next, which may suggest various other modifications.
In addition, I'll check the Android bugs and lint failures you raised. If we can't solve them on our side, I am afraid we will need to try unofficial fixes. In addition, I also need to ask you to confirm the version of flutter on your computer, check if it is on the latest stable version, because I see that your screenshot seems to have a strange version number.
Thanks :)

File: lib/generated/intl/messages_en.dart [hunk 21]
```diff
@@ -1,3 +1,5 @@
+// TODO: remove sdk version selector after migrating to null-safety.
+// @dart=2.10
```
Comment (Xanonymous-GitHub) [inline lib/generated/intl/messages_en.dart:2]: Suggest not adding anything into this auto-generated file.

Please help remove them by restart the intl generate process:
`flutter pub global run intl_utils:generate`

If you want to fix the import-non-null-safety warning from the analyzer, please add this comment
```dart
// ignore_for_file: import_of_legacy_library_into_null_safe
```
on the top of the file.

File: lib/generated/intl/messages_zh_TW.dart [hunk 22]
```diff
@@ -1,3 +1,5 @@
+// TODO: remove sdk version selector after migrating to null-safety.
+// @dart=2.10
```
Comment (Xanonymous-GitHub) [inline lib/generated/intl/messages_zh_TW.dart:2]: ditto.

File: lib/l10n/intl_zh_TW.arb [hunk 23]
```diff
@@ -30,7 +30,6 @@
   "accountLock": "帳號已被鎖住，請15分鐘後再試",
   "networkError": "網路發生錯誤",
   "unknownError": "發生未知錯誤",
-  "unknownServerError": "與校園系統溝通時發生問題，大部分功能可能將受影響\n請確認校園系統能正常使用後再試一次，謝謝",
```
Comment (Xanonymous-GitHub) [inline lib/l10n/intl_zh_TW.arb:33]: Please revert this change for removing, some of features depend on this resource.

File: pubspec.yaml [hunk 24]
```diff
@@ -167,7 +174,7 @@ flutter_intl:
   # Because it will be repeatedly generated when it is turned on,
   # it is very troublesome, so we temporarily turn it off.
   # If necessary, please turn it on by yourself.
-  enabled: false
+  enabled: true
```
Comment (Xanonymous-GitHub) [inline pubspec.yaml:177]: Please revert this change since we don't expect it to be set to `true`.

File: pubspec.yaml [hunk 25]
```diff
@@ -139,6 +143,9 @@ dependencies:
 
   weekday_selector: ^1.1.0
 
+  # 相機
```
Comment (Xanonymous-GitHub) [inline pubspec.yaml:146]: Suggest DON't use Chinese for comment, use English instead. The previous comments for other existing package does not plays a good example, don't follow them.
And also suggest not just repeat the package name since it is just a redundant comment without any meaningful descriptions.

File: pubspec.yaml [hunk 26]
```diff
@@ -123,6 +124,9 @@ dependencies:
   numberpicker: ^2.1.1
   #選擇學期
   alice: ^0.2.5
+  sensors_plus: ^1.3.1
+  #照片顯示
```
Comment (Xanonymous-GitHub) [inline pubspec.yaml:128]: ditto.

**Review by Xanonymous-GitHub — CHANGES_REQUESTED** [review]
Due to lack of time, I propose these revisions for the time being. In the next few days, I will conduct a full review and test of the remaining parts. Before that, please correct the following items, thank you.

More importantly, please carefully **split each revision or change into a different commit**, and write down the revision commit to each comment block after push, so that the reviewer can easily re-check it, thanks.

File: lib/generated/intl/messages_en.dart (hunk 21 above)
Comment (ggjj321) [reply lib/generated/intl/messages_en.dart]: solve commit : [22659b7](https://github.com/NEO-TAT/tat_flutter/pull/83/commits/22659b722eac53b1faab42e35b2ba3bf9a341877)

File: lib/generated/intl/messages_zh_TW.dart (hunk 22 above)
Comment (ggjj321) [reply lib/generated/intl/messages_zh_TW.dart]: solve commit : [0b067a5](https://github.com/NEO-TAT/tat_flutter/pull/83/commits/0b067a51628bdc6e70adbb01f7816f087e047913)

File: lib/l10n/intl_zh_TW.arb (hunk 23 above)
Comment (ggjj321) [reply lib/l10n/intl_zh_TW.arb]: resolve commit : [5afe7e5](https://github.com/NEO-TAT/tat_flutter/pull/83/commits/5afe7e5a83e28ed2f4ec3f52c7226082086d8169)

File: pubspec.yaml (hunk 24 above)
Comment (ggjj321) [reply pubspec.yaml]: resolve commit : [6f5fdf4](https://github.com/NEO-TAT/tat_flutter/pull/83/commits/6f5fdf4f1d2bfd694669e5e1bb855530e6c6ba2b)

File: pubspec.yaml (hunk 25 above)
Comment (ggjj321) [reply pubspec.yaml]: resolve commit : [11f249f](https://github.com/NEO-TAT/tat_flutter/pull/83/commits/11f249fe04736ad45ba7c7b20e48773baa558764)

File: pubspec.yaml (hunk 26 above)
Comment (ggjj321) [reply pubspec.yaml]: resolve commit : [728fbf](https://github.com/NEO-TAT/tat_flutter/pull/83/commits/f728fbf7142e1448c76531c89e70c3ee24a62933)

**Comment by ggjj321** [comment]
I want to ask how to solve the problem. I remember that I can't directly change this file.
image

**Comment by Xanonymous-GitHub** [comment]
@ggjj321 The solution has already provided in https://github.com/NEO-TAT/tat_flutter/pull/83#discussion_r985265418.

File: lib/ui/pages/coursedetail/screen/album_page.dart [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+
+  void _removePicture(picture) {
+    _pictureStorage.deletePicture(picture);
+    _pictures.remove(picture);
+    setState(() {});
+    Get.back();
+  }
+
+  void _useCourseIdToGetPicturePaths() async {
+    dynamic picturesInfo = await _pictureStorage.getCoursePicture(widget.courseId);
```
Comment (Xanonymous-GitHub) [inline lib/ui/pages/coursedetail/screen/album_page.dart:39]:
```suggestion
    final picturesInfo = await _pictureStorage.getCoursePicture(widget.courseId);
```

Suggest using `final` here.

File: lib/ui/pages/coursedetail/screen/album_page.dart [hunk 28]
```diff
@@ -0,0 +1,132 @@
...
+    for (final pictureInfo in picturesInfo) {
+      int infoId = pictureInfo['_id'];
+      String infoPath = pictureInfo['picturePath'];
+      String infoLabel = pictureInfo['label'];
+      String infoNote = pictureInfo['note'];
```
Comment (Xanonymous-GitHub) [inline lib/ui/pages/coursedetail/screen/album_page.dart:44]:
```suggestion
      final int infoId = pictureInfo['_id'];
      final String infoPath = pictureInfo['picturePath'];
      final String infoLabel = pictureInfo['label'];
      final String infoNote = pictureInfo['note'];
```

Suggest adding `final` to all available places.

File: lib/ui/pages/coursetable/course_table_page.dart [hunk 29]
```diff
@@ -656,13 +656,24 @@ class _CourseTablePageState extends State<CourseTablePage> {
...
+                  label: Text(R.current.details),
+                ),
               ]
             : [const SizedBox.shrink()],
       ),
       barrierDismissible: true,
     );
   }
 
+  Future<void> _showCameraPage(CourseInfoJson courseInfo) async {
```
Comment (Xanonymous-GitHub) [inline lib/ui/pages/coursetable/course_table_page.dart:671]:
```suggestion
  void _showCameraPage(CourseInfoJson courseInfo) {
```

The function seems not only not return a `Future` but also not uses any `await` keyword inside, so suggest make it return `void` and remove `async`.

**Review by Xanonymous-GitHub — COMMENTED** [review]
Hi @ggjj321 , for pushing the process of this PR, I've made some reviews about your style issues. Please check them first.

Once you clicked the `ready for review` button, I'll start to make a full review. Be careful to pass all CI checks and resolve all conflicts.

**Comment by ntut-xuan** [comment]
Since a long time ago (approach 1 year) and no additional commits to make the changes, I have decided to close the PR.
